examples/sparse/graphsage.py

```diff
@@ -0,0 +1,283 @@
+                mat = A.sample(1, fanout=node_num, ids=dst)
+                # Compact the matrix.
+                mat_cmp, src = mat.compact(0)
+                x = feat[src]
+                h = layer(mat_cmp, x)
+                if l != len(self.layers) - 1:
+                    h = F.relu(h)
+                    h = self.dropout(h)
+                y[dst] = h.to(buffer_device)
+            feat = y
+        return y
+
+
+def evaluate(model, dataloader, dataset, num_classes):
+    model.eval()
+    ys = []
+    y_hats = []
+    fanout = [10, 10, 10]
+    for it, (dst) in enumerate(dataloader):
+        with torch.no_grad():
+            src = dst
+            A_sample = []
+            for fout in fanout:
+                # Sampling neighbor
+                mat = A.sample(1, fout, ids=src, replace=True)
+                # Compact the matrix
+                mat_cmp, src_idx = mat.compact(0)
+                A_sample.append(mat_cmp)
+                src = src_idx
+
+            A_sample.reverse()
+            csrc = src.to("cpu")
+            cdst = dst.to("cpu")
+
+            x = dataset[0].ndata["feat"].index_select(0, csrc).to(device)
+            y = dataset[0].ndata["label"].index_select(0, cdst).to(device)
+            ys.append(y)
+            y_hats.append(model(A_sample, x))
+
+    return MF.accuracy(
+        torch.cat(y_hats),
+        torch.cat(ys),
+        task="multiclass",
+        num_classes=num_classes,
+    )
+
+
+def layerwise_infer(device, A, dataset, model, num_classes, batch_size):
+    model.eval()
+    nid = dataset.test_idx
+    with torch.no_grad():
+        pred = model.inference(
+            A, dataset, device, batch_size
+        )  # pred in buffer_device
+        pred = pred[nid]
+        label = dataset[0].ndata["label"][nid].to(pred.device)
+        return MF.accuracy(
+            pred, label, task="multiclass", num_classes=num_classes
+        )
+
+
+def train(device, A, dataset, model, num_classes):
+    # Create sampler & dataloader.
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_idx, batch_size=1024, shuffle=True
+    )
+    val_dataloader = torch.utils.data.DataLoader(val_idx, batch_size=1024)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanout = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (dst) in enumerate(train_dataloader):
+            # print(dst)
+            src = dst
+            A_sample = []
+            for fout in fanout:
+                # Sampling neighbor
+                mat = A.sample(1, fout, ids=src, replace=True)
+                # Compact the matrix
+                mat_cmp, src_idx = mat.compact(0)
+                A_sample.append(mat_cmp)
+                src = src_idx
+
+            A_sample.reverse()
+            csrc = src.to("cpu")
```

Comment:
Why move `src` and `dst` to CPU? The dataset should be moved to `device` before the index select.

Reply:
Ok.